File: 0347-39-3160-find-the-number-of-distinct-colors-among-the-balls/0347-39-3160-find-the-number-of-distinct-colors-among-the-balls.js
var queryResults = function(limit, queries) {
    let n = queries.length;
    let result = new Array(n);
    let ballColorMap = new Map();
    let colorCount = new Map();

    for (let i = 0; i < n; i++) {
        let [ball, color] = queries[i];

        if (ballColorMap.has(ball)) {
            let oldColor = ballColorMap.get(ball);
            colorCount.set(oldColor, colorCount.get(oldColor) - 1);
            if (colorCount.get(oldColor) === 0) colorCount.delete(oldColor);
        }

        colorCount.set(color, (colorCount.get(color) || 0) + 1);
        ballColorMap.set(ball, color);
        result[i] = colorCount.size;
    }

    return result;
};

File: 0347-39-3160-find-the-number-of-distinct-colors-among-the-balls/0347-39-3160-find-the-number-of-distinct-colors-among-the-balls.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        int n = queries.size();
        vector<int>v(n);
        unordered_map<int,int>mp,colorCnt;
        for(int i = 0 ; i < n ;i++){
            int ball = queries[i][0], color = queries[i][1];
            if(mp[ball]){
                int oldColor = mp[ball];
                colorCnt[oldColor]--;
                if(!colorCnt[oldColor])colorCnt.erase(oldColor);
            }
            colorCnt[color]++;
            mp[ball] = color;
            v[i] = (colorCnt.size());
        }
        return v;
    }
};
auto init = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 0;
}();